content/blog/opa-wrapper-plugin-intro/opa-wrapper-intro.md

```diff
@@ -0,0 +1,198 @@
+  AuthorizeResult,
+  PolicyDecision,
+  isPermission,
+} from '@backstage/plugin-permission-common';
+import {
+  catalogConditions,
+  createCatalogConditionalDecision,
+} from '@backstage/plugin-catalog-backend/alpha';
+import {
+  catalogEntityDeletePermission,
+} from '@backstage/plugin-catalog-common/alpha';
+
+class TestPermissionPolicy implements PermissionPolicy {
+  async handle(
+    request: PolicyQuery,
+    user?: BackstageIdentityResponse,
+   ): Promise<PolicyDecision> {
+    if (isPermission(request.permission, catalogEntityDeletePermission)) {
+      return createCatalogConditionalDecision(
+        request.permission,
+        catalogConditions.isEntityOwner({
+          claims: user?.identity.ownershipEntityRefs ?? [],
+        }),
+      );
+    }
+     return { result: AuthorizeResult.ALLOW };
+  }
+}
+```
+
+The above satisfies the following requirements:
+- If the permission is a `catalogEntityDeletePermission` then we want to check if the user is the owner of the entity and if they are then allow them to do the action, if they are not then deny them from doing the action.
+- If the permission is anything else then allow the user to do the action.
+
+In the above, if the conditions are met then the user will be allowed to do the action, if the conditions are not met then the user will be denied from doing the action, making this a conditional descision.
+
+So as you can see, the Permissions Framework is quite powerful and allows you to write some pretty complex policies and permissions. However, as I mentioned above, it does require you to write code and re-deploy your Backstage instance anytime you want to change your policies or permissions which may not be what you want to do.
+
+## OPA Permissions Wrapper Plugin 📦
+
+So now, again without going into too much detail as the technical stuff will come in a later post(s), lets take a quick look at how the OPA Permissions Wrapper Plugin works.
+
+We can use the same examples as above but instead of writing code we will write `.rego` files. Let's take a look at the first example:
+
+```rego
+package catalog_policy
+
+import future.keywords.if
+
+# Default decisions
+default allow := false
+
+# Don't allow deleting catalog entities
+allow if {
+ not input.permission.name = "catalog.entity.delete"
+}
+```
+
+The above policy in `.rego` is the same as the first example above. It will deny the user from doing the action if the permission is `catalog.entity.delete` and allow the user to do the action if the permission is anything else.
+
+Now lets take a look at the second example which is a conditional descision and slightly more complex:
+
+```rego
+package catalog_policy
+
+import future.keywords.if
+import future.keywords.in
+
+# Default decisions
+default allow := true
+
+default conditional := false
+
+# Set conditional
+conditional if {
+	is_delete
+}
+
+# condition for deleting
+condition := {"allOf": [{
+	"resourceType": "catalog-entity",
+	"rule": "IS_ENTITY_OWNER",
```

Comment:
This example was clearer to me when I read it in the post than in the code.

How I see it is that there are three logic steps:

* should enforce policy
* ok, hand over to OPA
* OPA says, ahh it's delete, then you need to enforce this other JSON policy
* backstage permissions enforces the JSON policy

I think that this is ok, I also think that your intuition to include those template JSON policies elsewhere is a good move. However, I does get me wondering? Is there a way that we can provide OPA with enough data to determine if the user is the entity owner?

It might be that this is not possible or impractical, but I think that'd be more idiomatic.

Reply:
> This example was clearer to me when I read it in the post than in the code.
> 
> How I see it is that there are three logic steps:
> 
> * should enforce policy
> * ok, hand over to OPA
> * OPA says, ahh it's delete, then you need to enforce this other JSON policy
> * backstage permissions enforces the JSON policy
> 
> I think that this is ok, I also think that your intuition to include those template JSON policies elsewhere is a good move. However, I does get me wondering? Is there a way that we can provide OPA with enough data to determine if the user is the entity owner?
> 
> It might be that this is not possible or impractical, but I think that'd be more idiomatic.

Yeah! You are on the right track indeed, sadly because I am wrapping around the Permission Framework all the information I am sending to OPA is all the information I can send (I _think_). So just to re-iterate so we're on the same page:

`Backstage` sends the input to OPA which usually contains a `permission` on an `entity` (let's say the catalog) which is this page: https://demo.backstage.io/catalog?filters%5Bkind%5D=component&filters%5Buser%5D=owned and a user/group, I then in `OPA` build either a `rule` or `policy` saying something like the group `readonly` should only be able to read. So the input would look like this:

```json
{
    "identity": {
        "groups": [
            "user:default/parsifal-m",
            "group:default/readonly"
        ],
        "username": "user:default/parsifal-m"
    },
    "permission": {
        "action": "read",
        "name": "catalog.entity.read",
        "resourceType": "catalog-entity",
        "type": "resource"
    }
}
```

Then I can either make a `conditional` rule or just return `allow` or `deny` if I want it to be conditional, lets say (in the link I sent) I only want you to be able to `read` components or APIs then I need OPA to also return the condition in the response so Backstage can apply that condition, i.e its `allowed` so long as this condition is met.

Which happens here:

https://github.com/Parsifal-M/brewed-backstage/blob/e7adc53c7625ab90f7807d083ffdc0e25938dba9/plugins/opa-permissions-wrapper/src/opa-evaluator/opaPermissionEvaluator.ts#L42-L63

`createCatalogConditionalDecision` expects a permission which in this case is the `catalog-entity` as a whole and a condition which is what is returned by `OPA`

This was all pretty hard (but fun) because I am sort of limited here: https://backstage.io/docs/reference/plugin-permission-node.permissionpolicy.handle 

And the permission framework in Backstage works a little funky 😄 

It is also possible I just don't know enough yet or have not looked at it from another angle 😆 I did raise this with the maintainers but apparently all of this is untrodden land 😅 

Hope the above makes sense 😵

Reply:
Thanks for sharing the example input! I guess that in order to have more data available at policy evaluation time, you'd need an additional component to sync data into OPA.

So I suppose that unless you want to store/represent all the policy data in the Rego code or static JSON, then it might be required to use the conditional allows.

I think that having OPA only allow/deny would make this a lot cleaner, however, it might be that the conditional permissions give you more flexibility.

Like you mentioned elsewhere, it'd be possible to store the JSON conditional permission somewhere for authors to reference it, or expose it via some function which generates the backstage conditional JSON.